### DISCUSSION TITLE
tutorials/performance/using_servers

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#114] [User-contributed notes]
# tutorials/performance/using_servers

Engines like Godot provide increased ease of use thanks to their high level constructs and features. Most of them are accessed and used via the Scene System. Using nodes and resources simplifies pr...

https://docs.godotengine.org/en/stable/tutorials/performance/using_servers.html

**Comment by felipebottega** [comment]
In the last code: First, there should be an "extends .." at the top. Second, in the _body_moved function, the canvas_item variable wasn't declared anywhere and will cause an error. Finally, they wrote "Physics2DServer", but I believe it should be "PhysicsServer2D".

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11685

**Comment by XandaPanda42** [comment]
Does anyone have any more information about how RID's work with the servers? Getting some inconsistent behavior and I can't find enough information in the docs to figure out what's I'm doing wrong. Any more info would be greatly appreciated, especially lower level stuff about how it works. (I've looked at the source, but I don't speak C++ sadly.)

This page and the RID page basically say "Don't forget to free your RID when you're done using it." It also says they're pointers to the resource itself. 

But can two RID's with different values point to the same resource? If not, is there any actual difference between `rid_a = rid_b` and `rid_a = RID(rid_b)` 

Do resources free their own RID when the resource gets freed? Does that mean I only need to call `RenderingServer.free_rid()` on RID's that I created using the `RenderingServer.*_create()` methods, not ones taken from a resource with `get_rid()` or `World3D.scenario` etc?

I've been assuming I needed to free any RID I use when I'm done with it.

**Reply by clayjohn** [reply]
Think of RIDs as an integer offset into an array (that's not what they are, but it helps to illustrate how you should treat them).

They are integers, so `rid_a = rid_b` and `rid_a = RID(rid_b)` are indeed equivalent. Two RIDs with different values cannot point to the same resource. 

When you free an RID, you are asking the engine to free the underlying resource. The RID itself is not an object that needs to be freed. It is just an integer

> Do resources free their own RID when the resource gets freed?

Yes

Any time Godot creates an RID, it will handle freeing the RID. Therefore you need to be very careful to avoid copying the RID from a Godot-created resource, then trying to use it after that resource is freed. That is a dangling pointer effectively. It won't crash (which is why we use RIDs and not pointers), but your game won't work correctly. 

> Does that mean I only need to call RenderingServer.free_rid() on RID's that I created using the RenderingServer.*_create() methods, not ones taken from a resource with get_rid() or World3D.scenario etc?

Yes

> I've been assuming I needed to free any RID I use when I'm done with it.

If you free RIDs for resources that you don't own, you are going to have a very bad time

**Reply by XandaPanda42** [reply]
Thank you for the answer, sorry for the late response.

> If you free RIDs for resources that you don't own, you are going to have a very bad time

If I recall, I did indeed have a very bad time. Half the bugs I was encountering were because of this haha. 

So I think the main takeaway is that we only need to manage RIDs for objects we create. If you created a Resource on the server and were given an RID in return, you must free that RID when you're no longer using it. 

Appreciate the help :-D

**Comment by Calfur** [comment]
I think the page would benefit from some examples regarding how to clean up RIDs.

I have created a PR which modifies the last example to get rid of the memory leaks it has right now:
https://github.com/godotengine/godot-docs/pull/11750
